### ISSUE TITLE
C extension compilation may not work with --enable-frozen-string-literal

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#1786]
https://travis-ci.org/puma/puma/jobs/606523352

```
$ export RUBYOPT="--enable-frozen-string-literal"
...
current directory:
/home/travis/build/puma/puma/vendor/bundle/truffleruby/19.2.0.1/gems/jaro_winkler-1.5.4/ext/jaro_winkler
/home/travis/.rvm/rubies/truffleruby-19.2.0.1/bin/truffleruby -I
/home/travis/.rvm/rubies/truffleruby-19.2.0.1/lib/mri -r
./siteconf20191102-9701-lu7lpn.rb extconf.rb
/home/travis/.rvm/rubies/truffleruby-19.2.0.1/lib/truffle/rbconfig.rb:215:in
`replace': can't modify frozen String (FrozenError)
from
/home/travis/.rvm/rubies/truffleruby-19.2.0.1/lib/truffle/rbconfig.rb:215:in
`expand'
from /home/travis/.rvm/rubies/truffleruby-19.2.0.1/lib/mri/mkmf.rb:2684:in
`MakeMakefile'
from /home/travis/.rvm/rubies/truffleruby-19.2.0.1/lib/mri/mkmf.rb:62:in `<top
(required)>'
	from extconf.rb:3:in `require'
	from extconf.rb:3:in `require'
	from extconf.rb:3:in `<main>'
```

I haven't dug any deeper than that, but I'm guessing something in our modifications to `mkmf` or `rbconfig` or something like that may not work with frozen string literals?

@nateberkopec

**Comment by eregon** [comment]
Thank you for the bug report.
I'm a little curious about the reason for using `--enable-frozen-string-literal`, as that is likely to break many things, is hard to test for Ruby implementations, and the only way to protect files expecting mutable strings is for the files to use
```ruby
# frozen_string_literal: false
```
I don't mind adding those for core & stdlib in TruffleRuby, but I'm not sure how gems authors would feel about it. In any case, since that should solve this case, let's do it.

**Comment by eregon** [comment]
`ruby --enable-frozen-string-literal -rmkmf -e0` is enough to reproduce this bug.

**Comment by chrisseaton** [comment]
> is hard to test for Ruby implementations

Do we need to start testing all our core basic functionality like building C extensions with this flag set, or something like that, do you think?

**Comment by nateberkopec** [comment]
I added it because of the possibility that it may, someday, become the default.

**Comment by eregon** [comment]
> I added it because of the possibility that it may, someday, become the default.

Right, it makes sense to ensure Puma is compatible with that flag.

**Comment by aardvark179** [comment]
This is a pretty simple fix. `RbConfig.expand` replaces the existing string value, and we're normally careful to only call it with a duplicated string. Unfortunately we're missing one case.

**Comment by eregon** [comment]
This should be fixed by d1cf59f731a5a152b7865adca898789732b0a6d4.
